File: server/tools/get_headlines.js
// Skill: a handful of current headlines, from Google News' public RSS feed
// (no account, no key). Parsed with a small regex rather than pulling in an
// XML library — RSS <title> tags are simple and consistent enough that a
// full parser would be overkill for "give me a few headlines". Also used
// directly by the morning briefing composer.

const FEED_URL = 'https://news.google.com/rss?hl=en-US&gl=US&ceid=US:en';

function decodeEntities(text) {
  return text
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

function extractTitles(xml, limit) {
  const titles = [];
  const re = /<item>[\s\S]*?<title>([\s\S]*?)<\/title>/g;
  let match;
  while ((match = re.exec(xml)) && titles.length < limit) {
    const raw = match[1].replace(/^<!\[CDATA\[|\]\]>$/g, '').trim();
    if (raw) titles.push(decodeEntities(raw));
  }
  return titles;
}

export default {
  name: 'get_headlines',
  description: "Get a few current top news headlines. Use this when the user asks what's in the news, or for a news update.",
  parameters: {
    type: 'object',
    properties: {
      topic: {
        type: 'string',
        description: 'Optional topic to focus on, e.g. "technology" or "sports". Leave out for general top headlines.',
      },
      count: {
        type: 'number',
        description: 'How many headlines to return. Default 5.',
      },
    },
    required: [],
  },
  async run({ topic, count } = {}) {
    const limit = Math.max(1, Math.min(10, Number(count) || 5));
    const url = topic
      ? `https://news.google.com/rss/search?q=${encodeURIComponent(topic)}&hl=en-US&gl=US&ceid=US:en`
      : FEED_URL;

    try {
      const res = await fetch(url);
      if (!res.ok) return { ok: false, error: 'The news service is unavailable right now.' };
      const xml = await res.text();
      const headlines = extractTitles(xml, limit);
      if (!headlines.length) return { ok: false, error: 'No headlines found.' };
      return { ok: true, topic: topic || 'top stories', headlines };
    } catch {
      return { ok: false, error: "Couldn't reach the news service." };
    }
  },
};

File: backend/jarvis/tools/get_headlines.py
"""Top news headlines, from Google News' public RSS feed."""

from __future__ import annotations

import re
import xml.etree.ElementTree as ET
from urllib.parse import quote

from ..capabilities import CapabilitySpec, Risk
from ._http import get_text

FEED_URL = "https://news.google.com/rss?hl=en-US&gl=US&ceid=US:en"


def _titles(xml: str, limit: int) -> list[str]:
    try:
        root = ET.fromstring(xml)
    except ET.ParseError:
        return []
    out = []
    for item in root.iter("item"):
        title = (item.findtext("title") or "").strip()
        if title:
            out.append(re.sub(r"\s+", " ", title))
        if len(out) >= limit:
            break
    return out


def _run(topic: str | None = None, count: int = 5) -> dict:
    limit = max(1, min(10, int(count or 5)))
    url = (f"https://news.google.com/rss/search?q={quote(topic)}&hl=en-US&gl=US&ceid=US:en"
           if topic else FEED_URL)
    try:
        headlines = _titles(get_text(url), limit)
    except Exception:  # noqa: BLE001
        return {"ok": False, "error": "Couldn't reach the news service."}
    if not headlines:
        return {"ok": False, "error": "No headlines found."}
    return {"ok": True, "topic": topic or "top stories", "headlines": headlines}


SPEC = CapabilitySpec(
    id="builtin.get_headlines",
    name="get_headlines",
    description=("Get current news headlines, optionally about a topic. Use this when the user "
                 "asks what's happening or wants the news."),
    input_schema={"type": "object", "properties": {
        "topic": {"type": "string", "description": 'Optional topic, e.g. "space" or "Arsenal".'},
        "count": {"type": "integer", "description": "How many headlines. Default 5, max 10."}},
        "required": []},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=20.0,
)
